lib/rom/support/data_proxy.rb

```diff
@@ -0,0 +1,51 @@
+module ROM
+  module DataProxy
+    attr_reader :data, :header, :tuple_proc
+
+    NON_FORWARDABLE = [
+      :each, :to_a, :to_ary, :kind_of?, :instance_of?, :is_a?
+    ].freeze
+
+    def self.included(klass)
+      klass.class_eval do
+        extend ClassMethods
+        include Equalizer.new(:data)
+      end
+    end
+
+    def initialize(data, header, tuple_proc = self.class.tuple_proc)
+      @data = data
+      @header = header
+      @tuple_proc = tuple_proc
+    end
+
+    def each
+      return to_enum unless block_given?
+      data.each { |tuple| yield(tuple_proc[tuple]) }
```

Comment:
You know this makes me happy. As I said yesterday, this is what I would expect: the proc applied to the tuple on demand rather than all at once. I suppose then it will play nice with usages like:

```ruby
my_dataset.each.take(5) { |row| do_stuff }
```

So that the proc is only applied the 5 times it's needed.